### PR TITLE
enterprise data

### PR DESCRIPTION
**Pull request by adeelq**

### CODE REVIEW

File: Controller/Adminhtml/Run/Automapdatafields.php [hunk 1]
```diff
@@ -69,6 +69,7 @@ public function execute()
         } else {
             // get all possible datatifileds
             $datafields = $this->datafield->getContactDatafields();
+            $eeFields = $this->datafield->contactEnterpriseDataFields;
```
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: getContactDatafields returns EE data fields if available. why do you need $eeFields explicitly?

File: Controller/Adminhtml/Run/Automapdatafields.php [hunk 2]
```diff
@@ -86,15 +87,23 @@ public function execute()
...
-                     * map the succesful created datafield
-                     */
-                    $this->data->saveConfigData(
-                        'connector_data_mapping/customer_data/' . $key,
-                        strtoupper($datafield['name']),
-                        $scope,
-                        $scopeId
-                    );
+
+                    //map the successfully created datafield
```
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: The logic is the same in the if and the else. can we remove the if/else and just have the logic?

File: Model/Connector/Datafield.php [hunk 3]
```diff
@@ -280,6 +280,52 @@ class Datafield
             ],
         ];
 
+    public $contactEnterpriseDataFields
```
Comment (simon-letch) [inline Model/Connector/Datafield.php]: Private if not needed externally?

File: Controller/Adminhtml/Run/Automapdatafields.php (hunk 1 above)
Comment (adeelq) [inline Controller/Adminhtml/Run/Automapdatafields.php]: this needed for if/else later on

File: Controller/Adminhtml/Run/Automapdatafields.php (hunk 2 above)
Comment (adeelq) [inline Controller/Adminhtml/Run/Automapdatafields.php]: Not the same. This is auto mapping. If key exist in  $eeFields then it is enterprise and enterprise config path is different

File: Model/Connector/Datafield.php (hunk 3 above)
Comment (adeelq) [inline Model/Connector/Datafield.php]: using this property in another class. so either can have this public. or make it private with a public function returning the private property?

File: Controller/Adminhtml/Run/Automapdatafields.php (hunk 2 above)
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: Aha, missed the config path being different. 

$configPath = isset($eeFields[$key]) ? 'connector_data_mapping/enterprise_data/' : 'connector_data_mapping/customer_data/';

$this->data->saveConfigData(
    $configPath . $key,
    strtoupper($datafield['name']),
    $scope,
    $scopeId
);

This reduces the number of lines of duplicated code and shows the intent clearly. Think it's a worthwhile improvement.

File: Controller/Adminhtml/Run/Automapdatafields.php (hunk 1 above)
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: It feels to me that the fact that a data field is enterprise only should be a property of the class. That information shouldn't be exposed by having a separate property to get 'enterprise' data fields.

Can we add a field isEnterprise to the datafield?

Then we can remove this line, make contactEnterpriseDataFields private to stop people potentially modifying them.

File: Helper/Data.php [hunk 4]
```diff
@@ -142,7 +142,7 @@ public function __construct(
         \Magento\Framework\App\Helper\Context $context,
         \Magento\Store\Model\StoreManagerInterface $storeManager,
         \Magento\Customer\Model\CustomerFactory $customerFactory,
-        \Magento\Framework\Module\ModuleListInterface $moduleListInterface,
+        \Magento\Framework\Module\FullModuleList $fullModuleList,
```
Comment (simon-letch) [inline Helper/Data.php]: Why move from the interface type to the concrete type?

Comment (adeelq) [inline Helper/Data.php]: \Magento\Framework\Module\ModuleListInterface is implemented by two classes \Magento\Framework\Module\FullModuleList and \Magento\Framework\Module\ModuleList. By default \Magento\Framework\Module\ModuleList is returned. I need \Magento\Framework\Module\FullModuleList which works for both our code and newer code

Comment (simon-letch) [inline Helper/Data.php]: should we be using di to specify which implementation we need?

Comment (adeelq) [inline Helper/Data.php]: we could but there is nothing wrong with using this approach..

Comment (adeelq) [inline Helper/Data.php]: see line https://github.com/dotmailer/dotmailer-magento2-extension/blob/testing/Helper/Data.php#L136 we could use interface

Comment (simon-letch) [inline Helper/Data.php]: sure and if we were changing that line i'd say we should be wiring it up in di too.
it feels to me that this is a step backwards? injecting concrete classes is bad right?

File: Controller/Adminhtml/Run/Automapdatafields.php [hunk 5]
```diff
@@ -69,6 +69,7 @@ public function execute()
         } else {
             // get all possible datatifileds
             $datafields = $this->datafield->getContactDatafields();
+            $eeFields = $this->datafield->getEnterpriseDataFields();
```
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: i still think this is bad and datafields should have an isEnterprise property on them

Comment (adeelq) [inline Controller/Adminhtml/Run/Automapdatafields.php]: what would be the advantage to that. we can add another value to each datafield array telling which one is enterprise and which one is not. we still need to check the value before choosing which oath is needed for config save. also the merged array is sent to DM api. So extra isEnterprise is not something DM is expending. 

if you are saying just adding class property to Datafield class then we will not be able to tell which datafield is EE so path can be chosen.

Advantage of get only ee datafields could be use anywhere in futrue where we only need ee datafields separated from normal data fields

File: Helper/Data.php (hunk 4 above)
Comment (adeelq) [inline Helper/Data.php]: no need to change it in two places. it is either DI or constructor. using concrete class is not bad. they still implement the same interface. 

we could leave the interface as defined before and add extra lines in DI. nothing will change will be same thing.

File: Controller/Adminhtml/Run/Automapdatafields.php (hunk 5 above)
Comment (simon-letch) [inline Controller/Adminhtml/Run/Automapdatafields.php]: this still doesn't sit right with me. datafields is leaking the fact that some fields are enterprise only through having to use a separate method. it's not clear that getContactDataFields includes enterprise. it relies on implementation knowledge to determine that.

adding the property will enable us to easily be able to figure out which is enterprise. we'd be able to remove this line. then check $datafield.isEnterprise in the ternary oeprator to determine $configPath.

YAGNI would suggest not looking to the future unless we need to. nothing else needs to use a set of predefined enterprise only datafields at the moment so reduce surface area by not exposing it.

agreed, we can't post that data to dotmailer. encapsulating all the knowledge about that datafield (is it really a datafield? or a predefinedDataField or defaultDataField or automappedDataField?) within the class itself seems like the correct way to do it. although it's potentially too late for that without significant rework.

Comment (adeelq) [inline Controller/Adminhtml/Run/Automapdatafields.php]: getContactDataFields function gets you all possible available datafields. It does what it says.
getEnterpriseDataFields function gets you only enterprise datafields. it does what it says.

**Pull request closed without merging**